engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala

```diff
@@ -1,55 +1,43 @@
+import scala.collection.immutable.StrictOptimizedMapOps
 import scala.collection.immutable.Map
-import scala.collection.immutable.MapOps
-import scala.collection.IterableFactory
-import scala.collection.immutable.Iterable
-import scala.collection.MapFactory
-import scala.annotation.unchecked.uncheckedVariance
-import scala.collection.mutable.Builder
+import scala.collection.Iterator
+
+final private class IrLazyImMap[K, V](
+  val underlying: java.util.Map[K, V]
+) extends Map[K, V]
+    with StrictOptimizedMapOps[K, V, Map, Map[K, V]] {
+  def removed(key: K) =
+    new scala.jdk.CollectionConverters.MapHasAsScala(underlying).asScala.toMap
+      .removed(key)
+
+  def updated[V1 >: V](key: K, value: V1) =
+    new scala.jdk.CollectionConverters.MapHasAsScala(underlying).asScala.toMap
+      .updated(key, value)
+
+  def get(key: K) = Option(underlying.get(key))
+
+  def iterator = new Iterator[(K, V)]() {
+    val keys             = underlying.entrySet().iterator()
+    def hasNext: Boolean = keys.hasNext()
+    def next(): (K, V) = {
+      val entry = keys.next()
+      (entry.getKey, entry.getValue)
+    }
+  }
+
+  override def size = underlying.size
+
+  // Has to override the `contains` method because the underlying implementation
+  // calls `get(key)`
+  override def contains(key: K): Boolean = {
+    underlying.containsKey(key)
+  }
 
-final private class IrLazyImMap[K, +V](
-  val underlying: scala.collection.Map[K, V],
-  val defaultValue: K => V
-) extends AbstractMap[K, V]
-    with MapOps[K, V, Map, IrLazyImMap[K, V]] {
-
-  def get(key: K): Option[V] = underlying.get(key)
-
-  override def default(key: K): V = defaultValue(key)
-
-  override def iterableFactory: IterableFactory[Iterable] =
-    underlying.iterableFactory
-
-  def iterator: Iterator[(K, V)] = underlying.iterator
-
-  override def isEmpty: Boolean = underlying.isEmpty
-
-  override def mapFactory: MapFactory[Map] = underlying.mapFactory
-
-  override def concat[V2 >: V](
-    xs: collection.IterableOnce[(K, V2)]
-  ): IrLazyImMap[K, V2] =
-    new IrLazyImMap(underlying.concat(xs), defaultValue)
-
-  def removed(key: K): IrLazyImMap[K, V] =
-    new IrLazyImMap[K, V](underlying.removed(key), defaultValue)
-
-  def updated[V1 >: V](key: K, value: V1): IrLazyImMap[K, V1] =
-    new IrLazyImMap[K, V1](underlying.updated(key, value), defaultValue)
-
-  override def empty: IrLazyImMap[K, V] =
-    new IrLazyImMap[K, V](underlying.empty, defaultValue)
-
-  override protected def fromSpecific(
-    coll: collection.IterableOnce[(K, V)] @uncheckedVariance
-  ): IrLazyImMap[K, V] =
-    new IrLazyImMap[K, V](mapFactory.from(coll), defaultValue)
+}
 
-  override protected def newSpecificBuilder
-    : Builder[(K, V), IrLazyImMap[K, V]] @uncheckedVariance =
-    Map.newBuilder.mapResult((p: Map[K, V]) =>
-      new IrLazyImMap[K, V](p, defaultValue)
-    )
+object IrLazyImMap {
+  def apply[K, V](map: java.util.Map[K, V]): Map[K, V] =
+    new IrLazyImMap[K, V](map)
```

Comment:
I guess I prefer to remove this and use `new IrLazyImMap(...)` at call sites. That's however a minor comment.

Reply:
Done in cac8910.